lib/Sema/CSDiagnostics.cpp

```diff
@@ -4264,6 +4264,29 @@ bool ThrowingFunctionConversionFailure::diagnoseAsError() {
 
 bool InOutConversionFailure::diagnoseAsError() {
   auto *anchor = getAnchor();
+  auto *locator = getLocator();
+  auto path = locator->getPath();
+
+  if (!path.empty() &&
+      path.back().getKind() == ConstraintLocator::FunctionArgument) {
+    if (auto argApplyInfo = getFunctionArgApplyInfo(locator)) {
+      emitDiagnostic(anchor->getLoc(), diag::cannot_convert_argument_value,
+          argApplyInfo->getArgType(), argApplyInfo->getParamType());
+    } else {
+      auto contextualType = getConstraintSystem().getContextualType();
```

Comment:
Under what circumstances does happen? I think it's only safe to check for contextual type/purpose for failures which have locator with `ContextualType` as one of the last elements (minus conversions), everything else would be unrelated.

Reply:
Ah I get it, the contextual type is a function type so it would be something like `<...> -> contextual type -> function {argument, result}`.

Reply:
I think it would be worth adding an assert here `assert(locator->findLast<LocatorPathElt::ContextualType>());`

Reply:
Yeah, inside `repairFailures` in the case of `ConstraintLocator::FunctionArgument`, we only create this fix if the previous path element is either `ConstraintLocator::ApplyArgToParam` (which is handled in the `if` body) or `ConstraintLocator::ContextualType`. I'll add an assert!